### ISSUE TITLE
Same tracking ID being assigned to two different objects

### ISSUE THREAD

**Issue by raj-kapil** [ultralytics/ultralytics#3824]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Detection, Tracking, Other

### Bug

I am running model.track over my video frames with the following 
```results = model.track(source=INPUT_VIDEO_PATH, conf=CONF, tracker=f"{HOME}/test.yaml", persist=True, stream=True, device=0, verbose=SHOW_LOGS)```. 
Majority of the time the tracking id keeps on incrementing with each new object but sometime the same ID is being assigned to two different objects. What could be the possible reason for this and how can I mitigate this issue. This is one of the rare issue that I am facing with my resultant videos.

Tracking config I am using -

tracker_type: bytetrack 
track_high_thresh: 0.5
track_low_thresh: 0.1
new_track_thresh: 0.6
track_buffer: 30 
match_thresh: 0.9

**Comment by xaristeidou** [comment]
How can you change the tracking arguments? Do you have to find the python file of the tracker and change it there or is it possible to change it through YOLO model?

**Comment by raj-kapil** [comment]
`tracker=f"{HOME}/test.yaml"`. We can pass a yaml file that contains the parameters that you can modify the tracker's working.

**Comment by xaristeidou** [comment]
Where can i find more info about the parameters that can be used for each tracker, do you know?

**Comment by raj-kapil** [comment]
I mean you can read about them either through GPT, tracker documentation, source code :)

**Comment by glenn-jocher** [comment]
@raj-kapil each tracker usually has its own set of parameters that determine how it behaves during the tracking process. Some general parameters across different trackers might include:

- **`tracker_type`**: Determines the type of tracker used. Some options might be KCF, TLD, MEDIANFLOW, or ByteTrack.
- **`track_high_thresh`**: Confidence threshold for tracker to consider detections as a valid track. 
- **`track_low_thresh`**: Confidence threshold below which a tracked object is considered as lost.
- **`new_track_thresh`**: Confidence threshold above which a new track is initialized.
- **`track_buffer`**: The maximum length of a tracked object's history.
- **`match_thresh`**: The threshold for matching detected objects and existing tracked objects.

Specific information about the parameters used by each tracker can usually be found in the tracker's documentation or by inspecting the tracker's source code. This will give you a better understanding of how the tracker works and what each parameter does.

For most situations, the defaults should work reasonably well, but tuning these parameters can improve performance, especially in situations with challenging conditions or specific requirements.

**Comment by Laughing-q** [comment]
@raj-kapil I tested track on the 5mins video with the same tracker cfg and script as yours, but it's not reproducible to me.
```python
from ultralytics import YOLO
import torch

model = YOLO("yolov8n.pt")
results = model.track(
    source="/home/laughing/Videos/test_person.mp4",
    classes=0,
    tracker="./test.yaml",
    persist=True,
    stream=True,
)
for result in results:
    plot_im = result.plot()
    id = result.boxes.id
    assert len(torch.unique(id)) == len(id), f"There are the same ids in {id}."
```
the logic here is `len(torch.unique(id))` would be less than `len(id)` if there are same ids and it would raise an error, but the track goes smoothly without any error.
![pic-selected-230720-1437-13](https://github.com/ultralytics/ultralytics/assets/61612323/c4db36f9-3e77-4318-9c1d-3ece6a0f5fa2)

**Comment by raj-kapil** [comment]
hey @Laughing-q , thanks for looking into this. I ran your code over my video but in each frame my unique id and the box id are same . `TORCH ID: tensor([1.]), ID : tensor([1.])`. even when I get a new object in my video.

The problem is that my tracker is only assigning a single ID i.e.,1 to two different objects. You can take a look here [re-assignment issue](https://youtu.be/GWDNwUlorSQ) my sample notebook that recreates this issue [sample kaggle notebook](https://www.kaggle.com/code/kapilraj/tracker-id-reassignment-issue/edit). The video that I am running on is [sample video](https://youtu.be/N_rx4ddLyOo).

**Comment by glenn-jocher** [comment]
@raj-kapil, thank you for providing additional information, including your video and the notebook. Having gone through your video, it indeed seems like the tracker is assigning the same ID to two objects, which shouldn't be the case.

The main reason could be the parameters set for the tracker. The parameters control various aspects of how the tracker functions, such as confidence thresholds for creating new tracks, updating existing tracks, etc. If the thresholds aren't set right, a newly discovered object might be assigned the same ID as a currently tracked object instead of getting a new one.

Based on the data you've provided, it seems like there might be room for optimizing the tracker parameters to make it better suited to your specific use case. Again, the exact parameter values required would depend on various aspects of your application, like the nature of the videos, the objects being tracked, etc.

So you might need to experiment with different combinations of parameter values to see which one works best for you. This debugging process can take a bit of time and requires a good understanding of what each parameter does and how they affect the tracking output.

Lastly, It would also be beneficial to ensure the objects in your video are detected correctly by the model before being passed on to the tracker. If the model's detections are inconsistent or inaccurate, it could also lead to issues with tracking IDs.

**Comment by Laughing-q** [comment]
@raj-kapil ok I've tested your video with your cfg and I'm able to reproduce the issue. 
It's because bytetrack would keep caches for tracks, and in your case the first person left but the caches are still there, then in a second the second person came in and matched the track cache of the first one, so the id remains.

Modifying some cfg of bytetrack would help, like when I was testing your video, I modified the `track_buffer` from `30` to `10` then the second person got a new id.
